webapp/sources/rudder/rudder-web/src/main/elm/editor/sources/MethodElemUtils.elm

```diff
@@ -0,0 +1,64 @@
+module MethodElemUtils exposing (..)
+
+import DataTypes exposing (..)
+import List.Extra
+
+getAllCalls: MethodElem -> List MethodElem
+getAllCalls call =
+  call ::
+  case call of
+    Block _ b -> List.concatMap getAllCalls b.calls
+    _ -> []
+
+setIdRec : String -> List MethodElem -> (List MethodElem, Bool)
+setIdRec  newId elems =
+  case elems of
+    [] -> ([], False)
+    head :: tail ->
+      if ((getId head ) == CallId "") then
+        ((setId (CallId newId) head ) :: tail, True)
```

Comment:
ok, it's here that we generate block id!

Reply:
But I still don't understand what happens for existing techniques, or imported one without an id

Reply:
existing techniques are migrated by Rudder, When we parse technique json on migration, we support missing id field by generating and id in case this case https://github.com/Normation/rudder/blob/e821a7e044d5aad0afbbbe1fbfcae8941ae5c4b4/webapp/sources/rudder/rudder-rest/src/main/scala/com/normation/rudder/rest/RestExtractorService.scala#L1228

I need to look for imported techniques but either it fails if id are missing or it generates one UI side (my guess is an error)